common/ast/navigable.go

```diff
@@ -87,34 +101,146 @@ func AllMatcher() ExprMatcher {
+// MatchDescendants takes a NavigableExpr and ExprMatcher and produces a list of NavigableExpr values
+// matching the input criteria in post-order (bottom up).
 func MatchDescendants(expr NavigableExpr, matcher ExprMatcher) []NavigableExpr {
-	return matchListInternal([]NavigableExpr{expr}, matcher, true)
+	matches := []NavigableExpr{}
+	navVisitor := &baseVisitor{
+		visitExpr: func(e Expr) {
+			nav := e.(NavigableExpr)
+			if matcher(nav) {
+				matches = append(matches, nav)
+			}
+		},
+	}
+	visit(expr, navVisitor, postOrder, 0, 0)
+	return matches
 }
 
 // MatchSubset applies an ExprMatcher to a list of NavigableExpr values and their descendants, producing a
 // subset of NavigableExpr values which match.
 func MatchSubset(exprs []NavigableExpr, matcher ExprMatcher) []NavigableExpr {
-	visit := make([]NavigableExpr, len(exprs))
-	copy(visit, exprs)
-	return matchListInternal(visit, matcher, false)
+	matches := []NavigableExpr{}
+	navVisitor := &baseVisitor{
+		visitExpr: func(e Expr) {
+			nav := e.(NavigableExpr)
+			if matcher(nav) {
+				matches = append(matches, nav)
+			}
+		},
+	}
+	for _, expr := range exprs {
+		visit(expr, navVisitor, postOrder, 0, 1)
+	}
+	return matches
+}
+
+// Visitor defines an object for visiting Expr and EntryExpr nodes within an expression graph.
+type Visitor interface {
+	// VisitExpr visits the input expression.
+	VisitExpr(Expr)
+
+	// VisitEntryExpr visits the input entry expression, i.e. a struct field or map entry.
+	VisitEntryExpr(EntryExpr)
+}
+
+type baseVisitor struct {
+	visitExpr      func(Expr)
+	visitEntryExpr func(EntryExpr)
+}
+
+// VisitExpr visits the Expr if the internal expr visitor has been configured.
+func (v *baseVisitor) VisitExpr(e Expr) {
+	if v.visitExpr != nil {
+		v.visitExpr(e)
+	}
+}
+
+// VisitEntryExpr visits the entry if the internal expr entry visitor has been configured.
+func (v *baseVisitor) VisitEntryExpr(e EntryExpr) {
+	if v.visitEntryExpr != nil {
+		v.visitEntryExpr(e)
+	}
+}
+
+// NewExprVisitor creates a visitor which only visits expression nodes.
+func NewExprVisitor(v func(Expr)) Visitor {
+	return &baseVisitor{
+		visitExpr:      v,
+		visitEntryExpr: nil,
+	}
 }
 
-func matchListInternal(visit []NavigableExpr, matcher ExprMatcher, visitDescendants bool) []NavigableExpr {
-	var matched []NavigableExpr
-	for len(visit) != 0 {
-		e := visit[0]
-		if matcher(e) {
-			matched = append(matched, e)
+// PostOrderVisit walks the expression graph and calls the visitor in post-order (bottom-up).
+func PostOrderVisit(expr Expr, visitor Visitor) {
+	visit(expr, visitor, postOrder, 0, 0)
+}
+
+// PreOrderVisit walks the expression graph and calls the visitor in pre-order (top-down).
+func PreOrderVisit(expr Expr, visitor Visitor) {
+	visit(expr, visitor, preOrder, 0, 0)
+}
+
+type visitOrder int
+
+const (
+	preOrder = iota + 1
+	postOrder
+)
+
+func visit(expr Expr, visitor Visitor, order visitOrder, depth, maxDepth int) {
+	if maxDepth > 0 && depth == maxDepth {
```

Comment:
Thoughts on aborting on abnormally deep AST? [Something similar to this](https://github.com/google/cel-java/commit/26545aa8935427be712bafd3f0abb378f367f2f6#diff-ffb0c1526486c5e8de07dbaba6750bf21a7c2cf22b8bc8ebfba30f40005e88dcR95)

Reply:
I think I might want to set this as an option on the visitor. Java handles exceptions well, but go doesn't like panics, and I'm not sure I want to put an error return status on the visit. Let me think about this more and come back to it.

Reply:
I've added a TODO. We'll finalize the behavior before the next release tag.